resdk/resources/data.py

```diff
@@ -48,171 +45,132 @@ def __init__(self, resolwe, **model_data):
-            # (b) self._collections is False since data object is not in any collection
-            self._collections = False
+    @process.setter
+    def process(self, payload):
+        """Set process."""
+        self._resource_setter(payload, Process, '_process')
 
-            # Note: ``collections`` field will only be present on single data
-            # object endpoint (e.g. /api/data/<ID>), but not on list endpoint
-            # (e.g. /api/data?id=<ID>). Therefore, explicit call on ``self.api``
-            # is made instead of self.reslwe.data.get (which makes request on
-            # list view but requires that list of results is of length 1)
-            payload = self.api(self.id).get()
-
-            if payload['collections']:
-                self._collections = self.resolwe.collection.filter(id__in=payload['collections'])
+    @property
+    def descriptor_schema(self):
+        """Get descriptor schema."""
+        return self._descriptor_schema
 
-        return self._collections
+    @descriptor_schema.setter
+    def descriptor_schema(self, payload):
+        """Set descriptor schema."""
+        self._resource_setter(payload, DescriptorSchema, "_descriptor_schema")
 
     @property
     def sample(self):
-        """Get ``sample`` that object belongs to."""
-        if self.id is None:
-            raise ValueError('Instance must be saved before accessing `sample` attribute.')
-        if self._sample is None:
-            # self._sample is set to False. This enables distinction between:
-            # (a) self._sample is None since self.sample was never accessed
-            # (b) self._sample is False since data object is not in any sample
-            self._sample = False
-
-            # Note: ``entities`` field will only be present on single data
-            # object endpoint (e.g. /api/data/<ID>), but not on list endpoint
-            # (e.g. /api/data?id=<ID>). Therefore, explicit call on ``self.api``
-            # is made instead of self.reslwe.data.get (which makes request on
-            # list view but requires that list of results is of length 1)
-            payload = self.api(self.id).get()
-
-            if payload['entities']:
-                self._sample = self.resolwe.sample.get(id=payload['entities'][0])
-
-        if self._sample:
-            return self._sample
-
-    @property
-    def descriptor_schema(self):
-        """Return descriptor schema assigned to the data object."""
-        if self._descriptor_schema is None:
-            return None
-
-        if self._hydrated_descriptor_schema is None:
-            if isinstance(self._descriptor_schema, int):
-                query_filters = {'id': self._descriptor_schema}
-            else:
-                query_filters = {'slug': self._descriptor_schema}
+        """Get sample."""
+        if self._original_values.get('entity', None):
+            self._sample = Sample(resolwe=self.resolwe, **self._original_values['entity'])
```

Comment:
You can load data into a `Sample` object in setter method, there is no overhead in doing that.

Reply:
Consider something like:
`d = res.data.get(id=1)` and than `d.sample`
In such cases attribute setting works like this::

    for field_name in self.fields():
        setattr(self, field_name, payload.get(field_name, None))

The issue is that data for `sample` attribute is stored in `entity` key in the payload. We have two options: (1) as done now or (2) by defining something like `attribute_mapping = {'sample': 'entity'}` as class attribute and then::

    for field_name in self.fields():
        field_name = self.attribute_mapping.get(field_name, field_name)
        setattr(self, field_name, payload.get(field_name, None))

Do you prefer the second option?